bridge/github/export.go

```diff
@@ -0,0 +1,731 @@
+	"github.com/MichaelMure/git-bug/bug"
+	"github.com/MichaelMure/git-bug/cache"
+	"github.com/MichaelMure/git-bug/util/git"
+)
+
+var (
+	ErrMissingIdentityToken = errors.New("missing identity token")
+)
+
+// githubExporter implement the Exporter interface
+type githubExporter struct {
+	conf core.Configuration
+
+	// number of exported bugs
+	exportedBugs int
+
+	// number of exported labels
+	exportedLabels int
+
+	// export only bugs taged with one of these origins
+	onlyOrigins []string
+
+	// cache identities clients
+	identityClient map[string]*githubv4.Client
+
+	// map identities with their tokens
+	identityToken map[string]string
+
+	// github repository ID
+	repositoryID string
+
+	// cache identifiers used to speed up exporting operations
+	// cleared for each bug
+	cachedIDs map[string]string
+
+	// cache labels used to speed up exporting labels events
+	cachedLabels map[string]string
+}
+
+// Init .
+func (ge *githubExporter) Init(conf core.Configuration) error {
+	ge.conf = conf
+	//TODO: initialize with multiple tokens
+	ge.identityToken = make(map[string]string)
+	ge.identityClient = make(map[string]*githubv4.Client)
+	ge.cachedIDs = make(map[string]string)
+	ge.cachedLabels = make(map[string]string)
+	return nil
+}
+
+// allowOrigin verify that origin is allowed to get exported.
+// if the exporter was initialized with no specified origins, it will return
+// true for all origins
+func (ge *githubExporter) allowOrigin(origin string) bool {
+	if ge.onlyOrigins == nil {
+		return true
+	}
+
+	for _, o := range ge.onlyOrigins {
+		if origin == o {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getIdentityClient return an identity github api v4 client
+// if no client were found it will initilize it from the known tokens map and cache it for next it use
+func (ge *githubExporter) getIdentityClient(id string) (*githubv4.Client, error) {
+	client, ok := ge.identityClient[id]
+	if ok {
+		return client, nil
+	}
+
+	// get token
+	token, ok := ge.identityToken[id]
+	if !ok {
+		return nil, ErrMissingIdentityToken
+	}
+
+	// create client
+	client = buildClient(token)
+	// cache client
+	ge.identityClient[id] = client
+
+	return client, nil
+}
+
+// ExportAll export all event made by the current user to Github
+func (ge *githubExporter) ExportAll(repo *cache.RepoCache, since time.Time) error {
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return err
+	}
+
+	ge.identityToken[user.Id()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID, err = getRepositoryNodeID(
+		ge.conf[keyOwner],
+		ge.conf[keyProject],
+		ge.conf[keyToken],
+	)
+
+	if err != nil {
+		return err
+	}
+
+	allBugsIds := repo.AllBugsIds()
+bugLoop:
+	for _, id := range allBugsIds {
+		b, err := repo.ResolveBug(id)
+		if err != nil {
+			return err
+		}
+
+		snapshot := b.Snapshot()
+
+		// ignore issues created before since date
+		if snapshot.CreatedAt.Before(since) {
+			continue
+		}
+
+		for _, p := range snapshot.Participants {
```

Comment:
if you write a `func (snap *Snapshot) HasParticipant(id string) bool` or even `func (snap *Snapshot) HasAnyParticipant(ids ...string) bool`, this code will magically simplify